src/Features/CSharp/Portable/Completion/CompletionProviders/AwaitCompletionProvider.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.MakeMethodAsynchronous;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
+{
+    /// <summary>
+    /// A completion provider for offering <see langword="await"/> keyword.
+    /// This is implemented separately, not as a keyword recommender as it contains extra logic for making container method async.
+    /// </summary>
+    /// <remarks>
+    /// The container is made async if and only if the containing method is returning a Task-like type.
+    /// </remarks>
+    [ExportCompletionProvider(nameof(AwaitCompletionProvider), LanguageNames.CSharp)]
+    [Shared]
+    internal sealed class AwaitCompletionProvider : LSPCompletionProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public AwaitCompletionProvider()
+        {
+        }
+
+        public override ImmutableHashSet<char> TriggerCharacters => CompletionUtilities.CommonTriggerCharactersWithArgumentList;
+
+        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            var document = context.Document;
+            var position = context.Position;
+            var cancellationToken = context.CancellationToken;
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
I'm wondering if it worth using `ReuseExistingSpeculativeModelAsync` here to create the syntaxContext and only try to get full model when we need to figure out the return type. @CyrusNajmabadi

Reply:
@genlu The feature was a bit flaky when I tried to use `ReuseExistingSpeculativeModelAsync`.

Reply:
does GetEnclosingSymbol not work with the speculative model.  We'd def want to use a speculative model here if we can.

Reply:
i'd still like to know the scenario where ReuseExistingSpeculativeModelAsync doesn't work.  can you be specific about the starting user scenario that that has a problem with?

Reply:
question still applies.

Reply:
Note: this is likely a blocker for this PR.